pylint/checkers/dataclass_checker.py

```diff
@@ -0,0 +1,114 @@
+
+
+def _check_name_or_attrname_eq_to(
+    node: nodes.Name | nodes.Attribute, check_with: str
+) -> Any:
+    """Utility function to check either a Name/Attribute node's name/attrname with a
+    given string.
+    """
+    if isinstance(node, nodes.Name):
+        return node.name == check_with
+    return node.attrname == check_with
+
+
+class DataclassChecker(BaseChecker):
+    """Checker that detects invalid or problematic usage in dataclasses.
+
+    Checks for
+    * invalid-field-call
+    """
+
+    name = "dataclass"
+    msgs = {
+        "E3601": (
+            "Invalid usage of field(), %s",
+            "invalid-field-call",
+            "The dataclasses.field() specifier should only be used as the value of "
+            "an assignment within a dataclass, or within the make_dataclass() function.",
+        ),
+    }
+
+    @utils.only_required_for_messages("invalid-field-call")
+    def visit_call(self, node: nodes.Call) -> None:
+        self._check_invalid_field_call(node)
+
+    def _check_invalid_field_call(self, node: nodes.Call) -> None:
+        """Checks for correct usage of the dataclasses.field() specifier in
+        dataclasses or within the make_dataclass() function.
+
+        Emits message
+        when field() is detected to be used outside a class decorated with
+        @dataclass decorator and outside make_dataclass() function, or when it
+        is used improperly within a dataclass.
+        """
+        if not isinstance(node.func, (nodes.Name, nodes.Attribute)):
+            return
+        if not _check_name_or_attrname_eq_to(node.func, "field"):
+            return
+        inferred_func = utils.safe_infer(node.func)
+        if not (
+            isinstance(inferred_func, nodes.FunctionDef)
+            and _is_dataclasses_module(inferred_func.parent)
+        ):
+            return
+        parent_node = node.parent
+        while parent_node and not isinstance(parent_node, (nodes.ClassDef, nodes.Call)):
+            parent_node = parent_node.parent
+
+        if isinstance(parent_node, nodes.Call):
+            inferred_func = utils.safe_infer(parent_node.func)
+        if not (
+            (  # Usage outside dataclass annotated classes
+                isinstance(parent_node, nodes.ClassDef) and parent_node.is_dataclass
+            )
+            or (  # Usage outside make_dataclass
+                isinstance(parent_node, nodes.Call)
+                and parent_node.func.name == "make_dataclass"
+                and _is_dataclasses_module(inferred_func.parent)
+            )
+        ):
+            self.add_message(
+                "invalid-field-call",
+                node=node,
+                args=(
+                    "it should be used within a dataclass or the make_dataclass() function.",
+                ),
+                confidence=INFERENCE,
+            )
+            return
+        if isinstance(parent_node, nodes.Call):
```

Comment:
This feels really weird. Can we separate the check for `Call` and `ClassDef`. We would duplicate one `add_message` but that feels more logical than these `if` statements.

Reply:
I slightly refactored it, let me know what you think